File: 113.path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.ans = []

    def path(self, root, sum, li):
        if not root.left and not root.right:
            if sum == root.val:
                li.append(sum)
                self.ans.append(li)
            return
        if root.left:
            self.path(root.left, sum - root.val, li + [root.val])
        if root.right:
            self.path(root.right, sum - root.val, li + [root.val])

    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if root:
            self.path(root, sum, [])
        return self.ans;

File: 113.path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> ans;
    void pathSum(TreeNode* root, int sum, vector<int> path) {
        if (!root->left && !root->right) {
            if (sum == root->val) {
                path.push_back(sum);
                ans.push_back(path);
            }
            return;
        }
        path.push_back(root->val);
        if (root->left) {
            pathSum(root->left, sum - root->val, path);
        }
        if (root->right) {
            pathSum(root->right, sum - root->val, path);
        }
    }

    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if (root) {
            pathSum(root, sum, vector<int>());
        }
        return ans;
    }
};
